### ISSUE TITLE
Implement `splice`

### ISSUE THREAD

**Issue by sunfishcode** [sunfishcode/mustang#91]
As reported [here](https://github.com/sunfishcode/mustang/issues/22#issuecomment-1005809920).

**Comment by al8n** [comment]
Hi, I am trying to implement `splice`, `vmsplice`, and `tee`, but I am confused about the defination for these `splice` and `vmsplice`. I have seen the definition of the `sendfile` is 
```rust
pub(crate) fn sendfile(
    out_fd: BorrowedFd<'_>,
    in_fd: BorrowedFd<'_>,
    offset: Option<&mut u64>,
    count: usize,
) -> io::Result<usize>
```

My confusions are:
1. For `ssize_t splice(int fd_in, off64_t *off_in, int fd_out, off64_t *off_out, size_t len, unsigned int flags);`, `fd_in` and `fd_out`, one of them must refer to a pipe, so I do not know use which type to limit the requirements.
2. For `ssize_t vmsplice(int fd, const struct iovec *iov, size_t nr_segs, unsigned int flags);`, I am confused use which type to indicate `iov`.

Could someone give me some instructions? Thanks!

**Comment by Urgau** [comment]
> My confusions are:
> 
> 1. For `ssize_t splice(int fd_in, off64_t *off_in, int fd_out, off64_t *off_out, size_t len, unsigned int flags);`, `fd_in` and `fd_out`, one of them must refer to a pipe, so I do not know use which type to limit the requirements.

To my knowledge there is currently no "branded file descriptor", so it's currently not possible to limit the type requirements as much as the interface want's.

> 2. For `ssize_t vmsplice(int fd, const struct iovec *iov, size_t nr_segs, unsigned int flags);`, I am confused use which type to indicate `iov`.

The [`iov`](https://docs.rs/libc/latest/libc/struct.iovec.html) struct is defined in the `libc` crate as well as in [`linux-raw-sys`](https://github.com/sunfishcode/linux-raw-sys/blob/96410022762afc99d2099f0fb61b98e4e0a68c23/src/v2_6_32/x86_64/general.rs#L2550-L2555) so the definition is not a problem. For the actual user facing type I would suggest using the same type as [`writev`](https://docs.rs/rustix/latest/rustix/io/fn.writev.html): [`IoSlice`](https://docs.rs/rustix/latest/rustix/io/struct.IoSlice.html). You can see here there usage regarding [`libc`](https://github.com/bytecodealliance/rustix/blob/1b93fef189fc589583f27fbbfee3988e938bfccf/src/imp/libc/io/syscalls.rs#L103-L112) and [`linux raw`](https://github.com/bytecodealliance/rustix/blob/687fcdd01a9a4dc5a493d48c0f1eb732d161cd96/src/imp/linux_raw/syscalls.rs#L365-L389).

I hope this helps.

**Comment by bjorn3** [comment]
> 1. For ssize_t splice(int fd_in, off64_t *off_in, int fd_out, off64_t *off_out, size_t len, unsigned int flags);, fd_in and fd_out, one of them must refer to a pipe, so I do not know use which type to limit the requirements.

I think it is fine to allow non-pipe file descriptors. If neither is a pipe, you will simply get EINVAL. You would get that error anyway if both file descriptors point to the same pipe.

**Comment by sunfishcode** [comment]
`splice` is implemented in #151!
